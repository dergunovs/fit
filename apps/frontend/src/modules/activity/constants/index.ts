import { TActivityChartType } from 'fitness-tracker-contracts';

export const URL_ACTIVITY = '/activities';
export const URL_ACTIVITY_CREATE = `${URL_ACTIVITY}/create`;
export const URL_ACTIVITY_EDIT = `${URL_ACTIVITY}/edit`;

export const URL_ACTIVITY_ADMIN = '/activities_admin';
export const URL_ACTIVITY_ADMIN_EDIT = `${URL_ACTIVITY_ADMIN}/edit`;

export const ACTIVITY_LIST_HEADERS = [{ title: 'Создано' }, { title: 'Пользователь' }];

export const CHART_TYPES: { title: string; value: TActivityChartType }[] = [
  { title: 'Занятия', value: 'activity' },
  { title: 'Подходы', value: 'set' },
  { title: 'Повторы', value: 'repeat' },
  { title: 'Мышцы', value: 'muscle' },
];

export const ACTIVITY_STATISTICS_GAP = 30;
