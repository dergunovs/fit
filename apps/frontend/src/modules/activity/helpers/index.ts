import { IActivity, IExerciseDone, IExerciseStatistics } from 'fitness-tracker-contracts';
import { createTempId, formatDate, formatDuration, subtractDates } from 'mhz-helpers';
import { toast } from 'mhz-ui';

export function getPotentialActivityDuration(
  exercises: IExerciseDone[],
  exerciseStatistics: IExerciseStatistics[],
  averageRestPercent?: number
) {
  if (!averageRestPercent) return '-';

  const totalDuration = exercises.reduce((acc, exercise) => {
    const averageDuration =
      exerciseStatistics.find((choosenExericse) => choosenExericse._id === exercise.exercise?._id)?.averageDuration ||
      0;

    return acc + averageDuration * exercise.repeats;
  }, 0);

  const durationWithRest = Math.round(totalDuration / (1 - averageRestPercent / 100));

  return formatDuration(durationWithRest);
}

export function convertActivityCalendarEvents(activities?: IActivity[]) {
  return activities?.map((activity: IActivity) => {
    return {
      _id: activity._id,
      start: new Date(`${activity.dateCreated}`),
      end: new Date(`${activity.dateUpdated}`),
      title: '+',
      content: activity.exercises,
    };
  });
}

export function generateActivityExercises(exercisesDone: IExerciseDone[]): IExerciseDone[] {
  const activityExercises = exercisesDone.map((exercise) => {
    return {
      _id: createTempId(),
      exercise: {
        _id: exercise.exercise?._id || '',
        title: exercise.exercise?.title || '',
        muscleGroups: exercise.exercise?.muscleGroups || [],
        isWeights: exercise.exercise?.isWeights || false,
        isWeightsRequired: exercise.exercise?.isWeightsRequired || false,
      },
      repeats: exercise.repeats,
      weight: exercise.weight,
    };
  });

  return activityExercises?.length ? [...activityExercises] : [];
}

export function getToFailurePercent(exercises: IExerciseDone[]) {
  const allExercises = exercises.length;
  const toFailureExercises = exercises.filter((exercise) => exercise.isToFailure).length;

  return `${Math.floor((toFailureExercises / allExercises) * 100)}%`;
}

export function getRestPercent(exercises: IExerciseDone[], start?: Date | null | string, end?: Date | null | string) {
  const activityDuration = Number(subtractDates(end, start, true));
  const exercisesDuration = exercises.reduce((acc, current) => acc + (current.duration || 0), 0);

  return `${Math.floor(100 - (exercisesDuration / activityDuration) * 100)}%`;
}

export async function copyActivityToClipboard(
  exercises: IExerciseDone[],
  start?: Date | null | string,
  end?: Date | null | string
) {
  const textHeader = `${formatDate(start, 'ru')}, длительность: ${subtractDates(end, start)}
Подходы: ${exercises.length}, отказы: ${getToFailurePercent(exercises)}, отдых: ${getRestPercent(exercises, start, end)}.

${exercises
  .map((exercise, index) => {
    return `${index + 1}. ${exercise.exercise?.title} x${exercise.repeats} ${exercise.weight ? `${exercise.weight}кг` : ''} ${formatDuration(exercise.duration)} ${exercise.isToFailure ? 'ДО ОТКАЗА' : ''}\n`;
  })
  .join('')}`;

  await navigator.clipboard.writeText(textHeader);

  toast.success('Скопировано в буфер');
}
