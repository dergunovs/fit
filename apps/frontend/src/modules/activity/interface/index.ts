export interface IActivityCalendarEvent<T> {
  _id?: string;
  start: Date | null;
  end: Date | null;
  title: string;
  content: T[];
  color?: string;
}

export interface ICalendarDates {
  dateFrom: string;
  dateTo: string;
}

export interface ITimelineStep {
  left: number;
  right: number;
  type: 'rest' | 'exercise';
}
