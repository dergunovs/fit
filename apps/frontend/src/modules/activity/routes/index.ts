import {
  URL_ACTIVITY_CREATE,
  URL_ACTIVITY_EDIT,
  URL_ACTIVITY_ADMIN,
  URL_ACTIVITY_ADMIN_EDIT,
} from '@/activity/constants';

export const activityRoutes = [
  {
    path: URL_ACTIVITY_CREATE,
    name: 'ActivityCreate',
    component: () => import('@/activity/pages/ActivityCreatePage.vue'),
  },
  {
    path: `${URL_ACTIVITY_EDIT}/:activity`,
    name: 'ActivityPassing',
    component: () => import('@/activity/pages/ActivityPassingPage.vue'),
  },
  {
    path: URL_ACTIVITY_ADMIN,
    name: 'ActivityList',
    component: () => import('@/activity/pages/ActivityListPage.vue'),
    meta: { layout: 'admin' },
  },
  {
    path: `${URL_ACTIVITY_ADMIN_EDIT}/:activity`,
    name: 'ActivityEdit',
    component: () => import('@/activity/pages/ActivityEditPage.vue'),
    meta: { layout: 'admin' },
  },
];
