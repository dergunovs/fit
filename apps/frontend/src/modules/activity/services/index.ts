import { ComputedRef, Ref } from 'vue';
import {
  API_ACTIVITY,
  API_ACTIVITY_CALENDAR,
  API_ACTIVITY_LAST,
  API_ACTIVITY_STATISTICS,
  API_ACTIVITY_CHART,
  TActivityChartType,
  TGetActivitiesDTO,
  TGetActivitiesQueryDTO,
  TGetActivitiesCalendarDTO,
  TGetActivitiesCalendarQueryDTO,
  TGetActivitiesStatisticsDTO,
  TGetActivitiesStatisticsQueryDTO,
  TGetActivitiesChartDTO,
  TGetActivitiesChartQueryDTO,
  TGetActivityDTO,
  TGetActivityLastDTO,
  TPostActivityDTO,
  TPostActivityDataDTO,
  TUpdateActivityDTO,
  TUpdateActivityDataDTO,
  TDeleteActivityDTO,
} from 'fitness-tracker-contracts';
import { useMutation, useQuery, api } from 'mhz-helpers';

export const activityService = {
  getMany: (page: Ref<number>) => {
    return useQuery({
      queryKey: [API_ACTIVITY, page],
      queryFn: async () => {
        const params: TGetActivitiesQueryDTO = { page: page.value };

        const { data } = await api.get<TGetActivitiesDTO>(API_ACTIVITY, { params });

        return data;
      },
    });
  },

  getStatistics: (gap: number) => {
    return useQuery({
      queryKey: [API_ACTIVITY_STATISTICS],
      queryFn: async () => {
        const params: TGetActivitiesStatisticsQueryDTO = { gap };

        const { data } = await api.get<TGetActivitiesStatisticsDTO>(API_ACTIVITY_STATISTICS, { params });

        return data;
      },
    });
  },

  getCalendar: (options: object, dateFrom: Ref<string>, dateTo: Ref<string>) => {
    return useQuery({
      queryKey: [API_ACTIVITY_CALENDAR, dateFrom, dateTo],
      queryFn: async () => {
        const params: TGetActivitiesCalendarQueryDTO = { dateFrom: dateFrom.value, dateTo: dateTo.value };

        const { data } = await api.get<TGetActivitiesCalendarDTO>(API_ACTIVITY_CALENDAR, { params });

        return data;
      },
      ...options,
    });
  },

  getChart: (type: Ref<TActivityChartType>, isMonth: Ref<boolean>, locale: Ref<string>) => {
    return useQuery({
      queryKey: [API_ACTIVITY_CHART, type, isMonth, locale],
      queryFn: async () => {
        const params: TGetActivitiesChartQueryDTO = {
          type: type.value,
          month: isMonth.value.toString(),
          locale: locale.value,
        };

        const { data } = await api.get<TGetActivitiesChartDTO>(API_ACTIVITY_CHART, { params });

        return data;
      },
    });
  },

  getOne: (options: object, id?: ComputedRef<string | undefined>) => {
    return useQuery({
      queryKey: [API_ACTIVITY, id],
      queryFn: async () => {
        if (!id?.value) return null;

        const { data } = await api.get<TGetActivityDTO>(`${API_ACTIVITY}/${id.value}`);

        return data.data;
      },
      ...options,
    });
  },

  getLast: () => {
    return useQuery({
      queryKey: [API_ACTIVITY_LAST],
      queryFn: async () => {
        const { data } = await api.get<TGetActivityLastDTO>(API_ACTIVITY_LAST);

        return data.data;
      },
    });
  },

  create: (options: object) => {
    return useMutation({
      mutationKey: [API_ACTIVITY],
      mutationFn: async (formData: TPostActivityDataDTO) => {
        const { data } = await api.post<TPostActivityDTO>(API_ACTIVITY, formData);

        return data;
      },
      ...options,
    });
  },

  update: (options: object) => {
    return useMutation({
      mutationKey: [API_ACTIVITY],
      mutationFn: async (formData: TUpdateActivityDataDTO) => {
        const { data } = await api.patch<TUpdateActivityDTO>(`${API_ACTIVITY}/${formData._id}`, formData);

        return data;
      },
      ...options,
    });
  },

  delete: (options: object) => {
    return useMutation({
      mutationKey: [API_ACTIVITY, API_ACTIVITY_CALENDAR],
      mutationFn: async (id: string) => {
        const { data } = await api.delete<TDeleteActivityDTO>(`${API_ACTIVITY}/${id}`);

        return data;
      },
      ...options,
    });
  },
};
