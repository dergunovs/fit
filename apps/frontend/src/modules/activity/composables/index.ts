import { ref } from 'vue';

import { ICalendarDates } from '@/activity/interface';

export function useActivityCalendar() {
  const dateFrom = ref('');
  const dateTo = ref('');

  const isDatesReady = ref(false);

  function updateDates(dates: ICalendarDates) {
    isDatesReady.value = true;

    dateFrom.value = dates.dateFrom;
    dateTo.value = dates.dateTo;
  }

  return { dateFrom, dateTo, isDatesReady, updateDates };
}
