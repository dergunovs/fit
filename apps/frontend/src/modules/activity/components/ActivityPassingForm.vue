<template>
  <div>
    <UiFlex column>
      <ExercisePassingList
        v-if="props.activity.exercises?.length"
        :exercises="props.activity.exercises"
        :activeExerciseId="activeExerciseId"
        @start="startExercise"
        @stop="stopExercise"
        data-test="exercise-passing-list"
      />

      <div>
        <div>
          Занятие создано <span data-test="activity-start">{{ formatDateTime(props.activity.dateCreated, 'ru') }}</span
          >.
        </div>

        <div v-if="props.activity.dateUpdated">
          Занятие {{ formData.isDone ? 'закончено' : 'обновлено' }}
          <span data-test="activity-updated">{{ formatDateTime(props.activity.dateUpdated, 'ru') }}</span
          >.
        </div>
      </div>

      <UiButton
        @click="finishActivity"
        layout="secondary"
        :isDisabled="props.activity.isDone"
        data-test="activity-finish"
      >
        Завершить занятие досрочно
      </UiButton>
    </UiFlex>
  </div>
</template>

<script setup lang="ts">
import { ref, onMounted, onBeforeUnmount } from 'vue';
import { useRouter } from 'vue-router';
import { toast, UiButton, UiFlex } from 'mhz-ui';
import { formatDateTime, useQueryClient, clone } from 'mhz-helpers';
import {
  API_ACTIVITY,
  API_ACTIVITY_CHART,
  API_ACTIVITY_STATISTICS,
  IActivity,
  IExerciseDone,
} from 'fitness-tracker-contracts';

import ExercisePassingList from '@/exercise/components/ExercisePassingList.vue';

import { activityService } from '@/activity/services';
import { URL_HOME } from '@/common/constants';

interface IProps {
  activity: IActivity;
}

const props = defineProps<IProps>();

const formData = ref<IActivity>({
  exercises: [],
  dateCreated: undefined,
  dateUpdated: undefined,
  dateScheduled: undefined,
  isDone: false,
});

const activeExerciseId = ref<string>();

const router = useRouter();

const queryClient = useQueryClient();

const { mutate: mutateUpdate } = activityService.update({
  onSuccess: async () => {
    await queryClient.refetchQueries({ queryKey: [API_ACTIVITY] });
  },
});

function startExercise(id: string) {
  activeExerciseId.value = id;
}

function stopExercise(exerciseDone: IExerciseDone) {
  activeExerciseId.value = undefined;

  formData.value.exercises = formData.value.exercises?.map((exercise) => {
    if (exercise._id === exerciseDone._id) {
      return {
        ...exercise,
        isDone: true,
        duration: exerciseDone.duration,
        isToFailure: exerciseDone.isToFailure,
        repeats: exerciseDone.repeats,
        dateUpdated: new Date(),
      };
    }

    return exercise;
  });

  formData.value.isDone = !formData.value.exercises?.some((exercise) => !exercise.isDone);

  if (!formData.value.dateUpdated) {
    formData.value.dateCreated = new Date(new Date().getTime() - (exerciseDone.duration || 0) * 1000);
  }

  formData.value.dateUpdated = new Date();

  mutateUpdate(formData.value);

  if (formData.value.isDone) exitActivity();
}

function finishActivity() {
  formData.value.isDone = true;

  mutateUpdate(formData.value);

  exitActivity();
}

function exitActivity() {
  toast.success('Занятие закончено');

  setTimeout(async () => {
    await queryClient.refetchQueries({ queryKey: [API_ACTIVITY_STATISTICS] });
    await queryClient.refetchQueries({ queryKey: [API_ACTIVITY_CHART] });

    router.push(URL_HOME);
  }, 1000);
}

const screenLock = ref<WakeLockSentinel>();

onMounted(async () => {
  if (props.activity) formData.value = clone(props.activity);

  screenLock.value = await navigator.wakeLock?.request();
});

onBeforeUnmount(() => {
  screenLock.value?.release();
  screenLock.value = undefined;
});
</script>
