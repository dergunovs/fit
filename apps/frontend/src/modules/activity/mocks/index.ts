import { ref, computed } from 'vue';
import { vi } from 'vitest';
import { TDeleteActivityDTO, TGetActivitiesDTO } from 'fitness-tracker-contracts';

import * as activityComposables from '@/activity/composables';
import { activityService } from '@/activity/services';
import {
  ACTIVITIES_CALENDAR_FIXTURE,
  ACTIVITIES_FIXTURE,
  ACTIVITIES_STATISTICS_FIXTURE,
  ACTIVITY_FIXTURE,
  ACTIVITIES_CHART_FIXTURE,
} from '@/activity/fixtures';
import { mockMutationReply, mockQueryReply } from '@/common/mocks';
import { IOnSuccess } from '@/common/interface';

const dateFrom = ref('2024-11-24T21:00:00.000Z');
const dateTo = ref('2025-01-05T20:59:59.000Z');
const isDatesReady = ref(true);
const spyUpdateDates = vi.fn();

const spyGetActivity = vi.spyOn(activityService, 'getOne').mockReturnValue(mockQueryReply(ACTIVITY_FIXTURE));

const getActivitiesData: TGetActivitiesDTO = { data: ACTIVITIES_FIXTURE, total: ACTIVITIES_FIXTURE.length };

const spyGetActivities = vi
  .spyOn(activityService, 'getMany')
  .mockImplementation(() => mockQueryReply(getActivitiesData));

const spyDeleteActivity = vi.fn();

const onSuccess: IOnSuccess = {
  delete: undefined,
};

vi.spyOn(activityService, 'delete').mockImplementation((options: { onSuccess?: () => Promise<void> }) => {
  if (options.onSuccess) onSuccess.delete = options.onSuccess;

  return mockMutationReply<TDeleteActivityDTO, string>(spyDeleteActivity);
});

const spyUseActivityCalendar = vi.spyOn(activityComposables, 'useActivityCalendar').mockReturnValue({
  dateFrom,
  dateTo,
  isDatesReady,
  updateDates: spyUpdateDates,
});

const computedActivityCalendarEvents = computed(() =>
  activityComposables.convertActivityCalendarEvents(ACTIVITIES_CALENDAR_FIXTURE)
);

const spyComputedActivityCalendarEvents = vi
  .spyOn(activityComposables, 'computedActivityCalendarEvents')
  .mockImplementation(() => computedActivityCalendarEvents);

const spyGetActivitiesCalendar = vi
  .spyOn(activityService, 'getCalendar')
  .mockImplementation(() => mockQueryReply(ACTIVITIES_CALENDAR_FIXTURE));

const spyGetActivitiesStatistics = vi
  .spyOn(activityService, 'getStatistics')
  .mockImplementation(() => mockQueryReply(ACTIVITIES_STATISTICS_FIXTURE));

const spyGetActivitiesChart = vi
  .spyOn(activityService, 'getChart')
  .mockImplementation(() => mockQueryReply(ACTIVITIES_CHART_FIXTURE));

export {
  spyGetActivities,
  spyGetActivity,
  spyGetActivitiesStatistics,
  spyGetActivitiesChart,
  spyDeleteActivity,
  spyUseActivityCalendar,
  spyComputedActivityCalendarEvents,
  spyGetActivitiesCalendar,
  spyUpdateDates,
  dateFrom,
  dateTo,
  isDatesReady,
  computedActivityCalendarEvents,
  getActivitiesData,
  onSuccess,
};
