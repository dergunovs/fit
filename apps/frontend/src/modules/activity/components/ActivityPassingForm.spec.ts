import { nextTick } from 'vue';
import { describe, it, expect, beforeEach, afterEach, vi } from 'vitest';
import { VueWrapper, enableAutoUnmount } from '@vue/test-utils';
import { API_ACTIVITY, API_ACTIVITY_CHART, API_ACTIVITY_STATISTICS } from 'fitness-tracker-contracts';
import { formatDateTime, dataTest, wait } from 'mhz-helpers';

import ActivityPassingForm from './ActivityPassingForm.vue';
import ExerciseElementList from '@/exercise/components/ExerciseElementList.vue';
import ExerciseElementPassing from '@/exercise/components/ExerciseElementPassing.vue';

import { wrapperFactory } from '@/common/test';
import { ACTIVITY_FIXTURE_2 } from '@/activity/fixtures';
import { mockOnSuccess, spyUpdateActivity } from '@/activity/mocks';
import { spyRefetchQueries, spyRouterPush, spyToastSuccess } from '@/common/mocks';
import { URL_HOME } from '@/common/constants';
import { updateExercisesIndex } from '@/exercise/helpers';

const restTimer = dataTest('activity-passing-form-rest-timer');
const exerciseList = dataTest('activity-passing-form-exercise-list');
const exerciseElement = dataTest('activity-passing-form-exercise');
const activityStart = dataTest('activity-start');
const activityFinish = dataTest('activity-finish');

const exerciseElementListStub = {
  name: 'ExerciseElementList',
  template: `
          <div data-test="activity-passing-form-exercises">
            <div v-for="(exercise, index) in exercises" :key="exercise._id">
              <slot :exercise="exercise" :index="index"></slot>
            </div>
          </div>
        `,
  props: ['exercises', 'isPassing'],
};

let wrapper: VueWrapper<InstanceType<typeof ActivityPassingForm>>;

beforeEach(() => {
  wrapper = wrapperFactory(
    ActivityPassingForm,
    { activity: ACTIVITY_FIXTURE_2 },
    { ExerciseElementList: exerciseElementListStub }
  );
});

enableAutoUnmount(afterEach);

describe('ActivityPassingForm', async () => {
  it('exists', async () => {
    expect(wrapper.findComponent(ActivityPassingForm)).toBeTruthy();
  });

  it('matches snapshot', async () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('shows start time', async () => {
    expect(wrapper.find(activityStart).text()).toBe(formatDateTime(ACTIVITY_FIXTURE_2.dateCreated, 'ru'));
  });

  it('shows rest timer', async () => {
    expect(wrapper.find(restTimer).exists()).toBe(true);
  });

  it('finishes activity early', async () => {
    expect(spyUpdateActivity).toBeCalledTimes(0);
    expect(spyRefetchQueries).toBeCalledTimes(0);
    expect(spyToastSuccess).toBeCalledTimes(0);

    expect(wrapper.find(activityFinish).attributes('isdisabled')).toBe(ACTIVITY_FIXTURE_2.isDone.toString());

    await wrapper.find(activityFinish).trigger('click');

    expect(spyUpdateActivity).toBeCalledTimes(1);
    expect(spyUpdateActivity).toBeCalledWith({ ...ACTIVITY_FIXTURE_2, isDone: true });

    await mockOnSuccess.update?.();

    expect(spyToastSuccess).toBeCalledTimes(1);

    expect(wrapper.find(activityFinish).exists()).toBe(false);

    await wait(1001);

    expect(spyRefetchQueries).toBeCalledTimes(3);
    expect(spyRefetchQueries).toBeCalledWith({ queryKey: [API_ACTIVITY] });
    expect(spyRefetchQueries).toBeCalledWith({ queryKey: [API_ACTIVITY_STATISTICS] });
    expect(spyRefetchQueries).toBeCalledWith({ queryKey: [API_ACTIVITY_CHART] });

    expect(spyRouterPush).toBeCalledTimes(1);
    expect(spyRouterPush).toBeCalledWith(URL_HOME);
  });

  it('stops exercise', async () => {
    const currentExerciseIndex = ACTIVITY_FIXTURE_2.exercises.filter((exercise) => exercise.isDone).length;
    const exercise = ACTIVITY_FIXTURE_2.exercises[currentExerciseIndex];
    const fixedDate = new Date();

    vi.setSystemTime(fixedDate);

    expect(wrapper.find(restTimer).exists()).toBe(true);

    wrapper.findComponent<typeof ExerciseElementPassing>(exerciseElement).vm.$emit('start', exercise._id);

    await nextTick();

    expect(wrapper.find(restTimer).exists()).toBe(false);

    const exerciseDone = { _id: exercise._id };

    wrapper.findComponent<typeof ExerciseElementPassing>(exerciseElement).vm.$emit('stop', exerciseDone, 60);

    await nextTick();

    expect(spyUpdateActivity).toBeCalledTimes(1);

    const expectedExercises = ACTIVITY_FIXTURE_2.exercises.map((e) =>
      e._id === exercise._id ? { ...e, isDone: true, duration: 60, dateUpdated: fixedDate } : e
    );

    expect(spyUpdateActivity).toBeCalledWith({
      ...ACTIVITY_FIXTURE_2,
      exercises: expectedExercises,
      dateUpdated: fixedDate,
      isDone: true,
    });

    vi.useRealTimers();
  });

  it('handles deleteExercise and updateIndex events correctly', async () => {
    const initialExercises = ACTIVITY_FIXTURE_2.exercises;
    const exerciseToDelete = initialExercises[0]._id;
    const newIndex = 1;

    expect(wrapper.findComponent<typeof ExerciseElementList>(exerciseList).props('exercises')).toEqual(
      initialExercises
    );

    wrapper.findComponent<typeof ExerciseElementList>(exerciseList).vm.$emit('delete', exerciseToDelete);

    await nextTick();

    const afterDeleteExercises = initialExercises.filter((e) => e._id !== exerciseToDelete);

    expect(wrapper.findComponent<typeof ExerciseElementList>(exerciseList).props('exercises')).toEqual(
      afterDeleteExercises
    );

    wrapper.findComponent<typeof ExerciseElementList>(exerciseList).vm.$emit('setIndex', newIndex);

    await nextTick();

    const expectedAfterIndexUpdate = updateExercisesIndex(afterDeleteExercises, newIndex);

    expect(wrapper.findComponent<typeof ExerciseElementList>(exerciseList).props('exercises')).toEqual(
      expectedAfterIndexUpdate
    );
  });
});
