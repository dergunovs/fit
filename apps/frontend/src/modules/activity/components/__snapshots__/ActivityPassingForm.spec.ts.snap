// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ActivityPassingForm > matches snapshot 1`] = `
"<div>
  <div column="">
    <exercise-passing-list-stub exercises="[object Object],[object Object]" data-test="exercise-passing-list"></exercise-passing-list-stub>
    <div>
      <div> Занятие началось <span data-test="activity-start">23 дек. 2024 г., 11:29</span>. </div>
      <div> Занятие обновлено <span data-test="activity-updated">23 дек. 2024 г., 12:21</span>. </div>
    </div><button layout="secondary" isdisabled="false" data-test="activity-finish"> Завершить занятие досрочно </button>
  </div>
</div>"
`;
